src/sensai/tensor_model.py

```diff
@@ -0,0 +1,392 @@
+"""
+This module contains base classes for models that input and output tensors, for examples CNNs.
+The fitting and predictions will still be performed on data frames, like in VectorModel,
+but now it will be expected that all entries of the input data frame passed to the model are tensors of the same shape.
+Lists of scalars of the same lengths are also accepted. The same is expected of the ground truth data frames.
+Everything will work as well if the entries are just scalars but in this case it is recommended to use
+VectorModel instead.
+
+If we denote the shapes of entries in the dfs as inputTensorShape and outputTensorShape,
+the model will be fit on input tensors of shape (N_rows, N_inputColumns, inputTensorShape) and output tensors of
+shape (N_rows, N_outputColumns, outputTensorShape), where empty dimensions (e.g. for one-column data frames)
+will be stripped.
+"""
+
+
+import logging
+from abc import ABC, abstractmethod
+from typing import Optional, Tuple
+
+import numpy as np
+import pandas as pd
+
+from .util.pandas import extractArray
+from .vector_model import VectorRegressionModel, VectorClassificationModel, VectorModel
+
+log = logging.getLogger(__name__)
+# we set the default level to debug because it is often interesting for the user to receive
+# debug information about shapes as data frames get converted to arrays
+log.setLevel(logging.DEBUG)
+
+
+class InvalidShapeError(Exception): pass
+
+
+def _getDatapointShape(df: pd.DataFrame):
+    firstRowDf = df.iloc[:1]
+    # Note that the empty first dimension with N_Datapoints=1 is stripped by extractArray
+    return extractArray(firstRowDf).shape
+
+
+def _checkDfShape(df: pd.DataFrame, desiredShape: tuple):
+    datapointShape = _getDatapointShape(df)
+    if datapointShape != desiredShape:
+        raise InvalidShapeError(f"Wrong input shape for data point. Expected {desiredShape} but got {datapointShape}")
+
+
+# This has to be implemented as a mixin because there can be no functional common class for tensor models.
+# The reason is that actual implementations need to inherit from Vector-Regression/Classification-Model
+# (or duplicate a lot of code) and thus it is not possible to inherit from something like TensorModel(VectorModel)
+# without getting into a mess. Despite that, we want things to "be a TensorModel", hence this class.
```

Comment:
I had two reasons in mind for TensorModel base classes:

1) (the less important one) - Normal vector model evaluators will not work with tensor models (which are also VectorModels). So I thought having separate classes and evaluators for them would make things a bit easier. We could use stuff like isinstance(TensorModel) for catching user errors

2) Better user experience. An important use case for tensor models will be that users implement them. Currently, only two methods need to be implemented (fitting and predicting on arrays). If a vector model "had a TensorModel", users would need to implement TensorModel and wrap it within VectorModel by themselves, making it much less straightforward. 

I don't like the mixin too much myself but the second point was quite important for me. I also tried to keep factorization in mind (there is quite some code duplication here for that reason ;) ). I am absolutely open about suggestions for better structuring, maybe we could have a call about that?

Reply:
Your first reason is actually a very good reason why they should _not_ be VectorModels. This is in violation of the Liskov substitution principle: Not all Vector(Regression|Classification)Models can be used in places where the interface specifies Vector(Regression|Classification)Model. This is a design error (not a user error).

Your second reason is also not convincing. If there are some special cases where the the behavioural violation does not apply, then you can offer an elegant wrapper construction, e.g. `Vector...ModelFromSpecialTensorModel(tensorModel)` or `SpecialTensorModel.toVector...Model()`.
Perhaps it would also be a better user experience, if one could use the more low-level TensorModel without the interface restrictions of VectorModel (cf. my TorchModel class, which has no ties to VectorModel at all).
Think about moving the functionality that is now in the VectorModel hierarchy to base classes of the new TL classes. I think it would be a cleaner design.

Reply:
I have adjusted to the form we had discussed (removed the class variables, undid the renaming of InputOutputData by rebasing). I think, this PR can be merged now.

Evaluation of tensor models will follow soon, training on data sets at some point in the future as well